src/compiler/parser.ts

```diff
@@ -1353,12 +1353,35 @@ namespace ts {
                 || isLiteralPropertyName();
         }
 
-        function nextTokenCanFollowDefaultKeyword(): boolean {
+        /**
+         * For export default assignments like
+         *  "export default foo"
+         *  "export default async"
+         * no modifiers should be parsed.
+         */
+        function nextTokenCanFollowDefaultModifier(): boolean {
             nextToken();
-            return token() === SyntaxKind.ClassKeyword || token() === SyntaxKind.FunctionKeyword ||
-                token() === SyntaxKind.InterfaceKeyword ||
-                (token() === SyntaxKind.AbstractKeyword && lookAhead(nextTokenIsClassKeywordOnSameLine)) ||
-                (token() === SyntaxKind.AsyncKeyword && lookAhead(nextTokenIsFunctionKeywordOnSameLine));
+            switch (token()) {
+                case SyntaxKind.ClassKeyword:
+                case SyntaxKind.FunctionKeyword:
+                case SyntaxKind.EnumKeyword:
+                case SyntaxKind.NamespaceKeyword:
+                case SyntaxKind.ModuleKeyword:
+                case SyntaxKind.ConstKeyword:
+                    return true;
+
+                // The following keywords aren't reserved keywords and could be identifiers.
+                // We need to look at the next token to make sure these should be parsed as a modifier keyword,
+                // and not as an identifier. If they are followed by an identifier or a keyword on the same line,
+                // the current statement is not an export default assignment.
+                // See https://github.com/Microsoft/TypeScript/blob/master/doc/spec.md#221-reserved-words
+                case SyntaxKind.InterfaceKeyword:
+                case SyntaxKind.AbstractKeyword:
+                case SyntaxKind.DeclareKeyword:
+                case SyntaxKind.AsyncKeyword:
+                case SyntaxKind.TypeKeyword:
+                    return lookAhead(nextTokenIsIdentifierOrKeywordOnSameLine);

```

Comment:
@rbuckton This simplified lookahead seems to be enough; see https://github.com/Microsoft/TypeScript/pull/18628/commits/5113249e4e3e9da299e9dfe804c7545dd9922b2e#diff-cc2127059a7624c4bd840e11d106cb81